### PR TITLE
(WIP) Mail rewrite

### PR DESCRIPTION
**Pull request by melwil**

### CODE REVIEW

File: onlineguru/src/main/java/no/ntnu/online/onlineguru/plugin/plugins/github/GithubCallback.java [hunk 1]
```diff
@@ -36,17 +36,14 @@ public GithubCallback(Wand wand, Listeners listeners) {
...
 
     @Override
-    public Response serve(String uri, String method, Properties header, Properties parms, Properties files) {
-        // make payload to get IRCAnnounce
-        logger.debug("Request received from webserver.");
-
+    public Response serve(String uri, Method method,  Map<String, String> headers, Map<String, String> parms, Map<String, String> files) {
         GithubPayload payload = null;
-        if (method.equalsIgnoreCase("POST") && parms.containsKey("payload")) {
+        if (Method.POST.equals(method) && parms.containsKey("payload")) {
```
Comment (espenja) [inline onlineguru/src/main/java/no/ntnu/online/onlineguru/plugin/plugins/github/GithubCallback.java:41]: I would do the check the other way around here: method.equalsIgnoreCase(Method.POST), semantically more obvious what you're trying to do

Comment (melwil) [inline onlineguru/src/main/java/no/ntnu/online/onlineguru/plugin/plugins/github/GithubCallback.java:41]: This is the way it's done inside NanoHTTPD, hence why I did it this way around.

I can change it though.

File: onlineguru/src/main/java/no/ntnu/online/onlineguru/plugin/plugins/mailannouncer/MailCallback.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        if (Method.POST.equals(method) && parms.containsKey("payload")) {
+            try {
+                mail = gson.fromJson(parms.get("payload"), Mail.class);
+            } catch (JsonParseException e) {
+                logger.error("Failed to parse JSON from "+uri, e.getCause());
+                logger.error(parms.get("payload"));
+            }
+        }
+
+        System.out.println("1");
```
Comment (espenja) [inline onlineguru/src/main/java/no/ntnu/online/onlineguru/plugin/plugins/mailannouncer/MailCallback.java]: remove debug

File: onlineguru/src/main/java/no/ntnu/online/onlineguru/plugin/plugins/mailannouncer/MailCallback.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        if (mail != null) {
+            System.out.println("2");
```
Comment (espenja) [inline onlineguru/src/main/java/no/ntnu/online/onlineguru/plugin/plugins/mailannouncer/MailCallback.java]: remove debug

File: onlineguru/src/main/java/no/ntnu/online/onlineguru/plugin/plugins/mailannouncer/MailCallback.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+            String mailinglist = mail.getMailinglist();
+            if (listeners.containsKey(mailinglist)) {
+                System.out.println("3");
```
Comment (espenja) [inline onlineguru/src/main/java/no/ntnu/online/onlineguru/plugin/plugins/mailannouncer/MailCallback.java]: remove debug

File: onlineguru/src/main/java/no/ntnu/online/onlineguru/plugin/plugins/mailannouncer/listeners/Listeners.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
+package no.ntnu.online.onlineguru.plugin.plugins.mailannouncer.listeners;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * @author Håvard Slettvold
+ */
+public class Listeners {
```
Comment (espenja) [inline onlineguru/src/main/java/no/ntnu/online/onlineguru/plugin/plugins/mailannouncer/listeners/Listeners.java]: "Listeners" of what? Rename to MailCallbackListeners

File: onlineguru/src/main/java/no/ntnu/online/onlineguru/plugin/plugins/mailannouncer/listeners/CallbackListener.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+import no.ntnu.online.onlineguru.plugin.plugins.mailannouncer.MailCallback;
+import no.ntnu.online.onlineguru.plugin.plugins.mailannouncer.model.Mail;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * @author Håvard Slettvold
+ */
+public class CallbackListener {
```
Comment (espenja) [inline onlineguru/src/main/java/no/ntnu/online/onlineguru/plugin/plugins/mailannouncer/listeners/CallbackListener.java]: "CallbackListener" of what? Rename to MailCallbackListener

File: onlineguru/src/main/java/no/ntnu/online/onlineguru/plugin/plugins/mailannouncer/model/Mail.java [hunk 7]
```diff
@@ -0,0 +1,34 @@
+package no.ntnu.online.onlineguru.plugin.plugins.mailannouncer.model;
+
+/**
+ * @author Håvard Slettvold
+ */
+public class Mail {
+
+    private String to;
```
Comment (espenja) [inline onlineguru/src/main/java/no/ntnu/online/onlineguru/plugin/plugins/mailannouncer/model/Mail.java:8]: Instansiate variables to "" or null

Comment (melwil) [inline onlineguru/src/main/java/no/ntnu/online/onlineguru/plugin/plugins/mailannouncer/model/Mail.java:8]: It's not normal to instantiate objects which are created by JSON libs, and they will be null by default.

File: onlineguru/src/main/java/no/ntnu/online/onlineguru/service/services/webserver/Webserver.java [hunk 8]
```diff
@@ -3,32 +3,46 @@
...
-    public Response serve(String uri, String method, Properties header, Properties parms, Properties files) {
-        logger.debug(String.format("uri: %s, method: %s, header: %s, parms: %s, files: %s", uri, method, header, parms, files));
-        if (services.containsKey(uri)) {
-            return services.get(uri).serve(uri, method, header, parms, files);
+    public Response serve(IHTTPSession session) {
+        logger.debug("Incoming request - " + session.getMethod() + " '" + session.getUri() + "'");
+
+        Map<String, String> files = new HashMap<String, String>();
+        Method method = session.getMethod();
+        if (Method.PUT.equals(method) || Method.POST.equals(method)) {
```
Comment (espenja) [inline onlineguru/src/main/java/no/ntnu/online/onlineguru/service/services/webserver/Webserver.java:29]: Other way around again, I do not agree with they way NanoHTTP does this internally, it looks wrong. You're not checking if Method.POST is equal to something, you're checking if method is equal to Method.POST (yeah yeah same thing, but you get what I mean)

Comment (melwil) [inline onlineguru/src/main/java/no/ntnu/online/onlineguru/service/services/webserver/Webserver.java:29]: Actually they are enums, which makes it unnecessary to use .equals.
Doing it this way around eliminates potential nullpointers, though, but I see what you mean.

File: onlineguru/src/main/java/no/ntnu/online/onlineguru/plugin/plugins/mailannouncer/listeners/Listeners.java (hunk 5 above)
Comment (melwil) [inline onlineguru/src/main/java/no/ntnu/online/onlineguru/plugin/plugins/mailannouncer/listeners/Listeners.java]: It's deliberately kept short as it's just a wrapper class around a hashmap containing the CallbackListeners.

File: onlineguru/src/main/java/no/ntnu/online/onlineguru/plugin/plugins/mailannouncer/listeners/CallbackListener.java (hunk 6 above)
Comment (melwil) [inline onlineguru/src/main/java/no/ntnu/online/onlineguru/plugin/plugins/mailannouncer/listeners/CallbackListener.java]: Agreed.

File: onlineguru/src/main/java/no/ntnu/online/onlineguru/plugin/plugins/mailannouncer/listeners/Listeners.java (hunk 5 above)
Comment (espenja) [inline onlineguru/src/main/java/no/ntnu/online/onlineguru/plugin/plugins/mailannouncer/listeners/Listeners.java]: Java isn't supposed to be short :p Semantically it's better, and also while searching for files. /me searches for a file and types in "Listeners", 9001 results expected.

Comment (melwil) [inline onlineguru/src/main/java/no/ntnu/online/onlineguru/plugin/plugins/mailannouncer/listeners/Listeners.java]: Fine, I'll change the name.

File: onlineguru/src/main/java/no/ntnu/online/onlineguru/service/services/webserver/Webserver.java (hunk 8 above)
Comment (espenja) [inline onlineguru/src/main/java/no/ntnu/online/onlineguru/service/services/webserver/Webserver.java:29]: Did nazi Method method = session.getMethod(). Presumed it was a String, carry on.

**Pull request merged**